### PR TITLE
Add `write_json`, `read_json`, `register_json`, and `JsonFormat` to `CREATE EXTERNAL TABLE` functionality

### PR DESCRIPTION
**Pull request by matthewmturner**
# Which issue does this PR close?

Part of #1777 reading / registering JSON files

 # Rationale for this change

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by matthewmturner** [comment]
I plan on submitting PR to https://github.com/apache/arrow-testing to get JSON test data added to improve testing setup for this.

**Comment by alamb** [comment]
I agree we should push ahead as is and either check in some small json data directly into datafusion, perhaps in https://github.com/apache/arrow-datafusion/tree/master/datafusion/tests

It isn't clear to me the value of adding to `arrow-testing`

**Review by houqp — APPROVED** [review]
LGTM, and agree with what @alamb said

**Comment by matthewmturner** [comment]
I was just surprised that there was no standard dataset for json when there was for all the other file formats - so i thought there was value in standardizing that.  But its no big deal.  Ill do locally here.

**Review by yjshen — APPROVED** [review]
LGTM, Thanks @matthewmturner !

File: datafusion/tests/sql/json.rs [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use super::*;
+
```
Comment (alamb) [inline datafusion/tests/sql/json.rs:19]: these tests seem to be testing `avro` not `json` -- I wonder if you forgot to push a commit @matthewmturner ?

**Comment by matthewmturner** [comment]
I added `write_json` as well so everything could be tested in a round trip

**Comment by matthewmturner** [comment]
So the current error is when reading a json file with a single record with one field that is null.  the schema is not picked up correctly. i think this would have to be handled on the arrow side.  maybe i could add `#[should_panic]` for now?

File: datafusion/tests/sql/json.rs [hunk 2]
```diff
@@ -17,135 +17,76 @@
...
-        NdJsonReadOptions::default(),
-    )
-    .await
-    .unwrap();
-}
+const TEST_DATA_BASE: &str = "tests/jsons";
 
 #[tokio::test]
-async fn avro_query() {
+async fn json_query() {
```
Comment (alamb) [inline datafusion/tests/sql/json.rs:23]: 👍

**Comment by alamb** [comment]
> So the current error is when reading a json file with a single record with one field that is null. the schema is not picked up correctly. i think this would have to be handled on the arrow side. maybe i could add #[should_panic] for now?


I am not sure what the schema of a single record with one null field should be (and doesn't seem all that important of a use case). Adding `#should_panic `sounds good to me and maybe filing a ticket in arrow-rs with whatever is wrong?

File: datafusion/tests/sql/avro.rs [hunk 3]
```diff
@@ -133,7 +133,7 @@ async fn avro_single_nan_schema() {
 }
 
 #[tokio::test]
-async fn avro_explain() {
+async fn json_explain() {
```
Comment (alamb) [inline datafusion/tests/sql/avro.rs:136]: this should probably still be `avro_explain` I think

Comment (alamb) [reply datafusion/tests/sql/avro.rs:136]:
```suggestion
async fn avro_explain() {
```

**Comment by matthewmturner** [comment]
To clarify when i said one null field i meant a single column and record, with that record having a null value.  For example:

```
{"mycol": null}
```

I would expect the schema to have the field "mycol" and it be nullable.  But instead no fields were picked up.

**Comment by alamb** [comment]
> I would expect the schema to have the field "mycol" and it be nullable. But instead no fields were picked up.

I guess I don't know what type I would  expect `mycol` to be. `DataType::Null` might be accurate but not likely all that useful. 🤔

**Comment by alamb** [comment]
Thanks again @matthewmturner

**Pull request merged**